### ISSUE TITLE
calculateResolvedType() for MethodCallExpr.getscope().get()  is not responding

### ISSUE THREAD

**Issue by ghost** [javaparser/javaparser#3123]
I'm analyzing the callees inside a given method and using the following snippet to get the classes declaring the callees.
```java
public void visit(MethodCallExpr n, Void arg) {
    if (n.getScope().isPresent()) {
        System.out.println("Current method call expression is: " + n.toString());
        Expression expression = n.getScope().get();
        System.out.println("Expression: " + expression);
        System.out.println("Callee is defined in Class: " + expression.calculateResolvedType().describe());
    }
}
```

However, calculateResolvedType() is not responding when resolving the following method call statement.
```java
ByteBuf driverId = Unpooled.copyLong(driverIdPart1, driverIdPart2);
// Unpooled is defined in a third-party library: netty
```
The output is:
```
Current method call expression is: Unpooled.copyLong(driverIdPart1, driverIdPart2)
Expression: Unpooled

```
There aren't any warnings and errors.
All libraries have been included in the path, and everything is ok for other methodcallexprs, such as:
```java
LOGGER.warn("Config parameter " + getProtocolName() + ".timezone is deprecated")
TimeZone.getTimeZone(defaultTimeZone)
```
And I've checked the netty library and made sure it contains `Unpooled` class.

**Comment by ghost** [comment]
I've been waiting for half an hour and it still hasn't responded.

Any help is appreciated. Thanks in advance.

**Comment by jlerbsc** [comment]
Please can you provide us a simple test case to reproduce this issue?

**Comment by ghost** [comment]
@jlerbsc Thank you for your prompt reply.

Please find the test-case from my GitHub [repo](https://github.com/liuyuanOUC/test-case), and run the `MethodSolver.java` which analyzes the Traccar software.

You can get the following result:
![image](https://user-images.githubusercontent.com/27879832/109797891-06053100-7c55-11eb-8425-4cc624cafaf9.png)

The exception occurs in line 201 in `traccar-4.6/src/main/java/org/traccar/protocol/RuptelaProtocolDecoder.java` file.

If you have any questions, please let me know.

**Comment by jlerbsc** [comment]
There is no issue with this simple test case. Can you use this simple test case to reproduce your issue?

    @Test
    void issue3123() throws IOException {
        Path pathToJar = Paths.get("D:\\tmp\\test-case\\dataset\\traccar-4.6\\lib\\netty-all-4.1.39.Final.jar");
        ParserConfiguration config = new ParserConfiguration();
        CombinedTypeSolver cts = new CombinedTypeSolver();
        cts.add(new ReflectionTypeSolver(false));
        cts.add(new JarTypeSolver(pathToJar));
        config.setSymbolResolver(new JavaSymbolSolver(cts));
        StaticJavaParser.setConfiguration(config);

        String str = 
                "import io.netty.buffer.Unpooled;\n" +
                "public class Foo {\n" +
                "  void a() {\n" +
                "    Unpooled.copyLong(1L, 1L);\n" +
                "  }\n" +
                "}\n";
        CompilationUnit cu = StaticJavaParser.parse(str);
        MethodCallExpr mce = cu.findFirst(MethodCallExpr.class).get();
        System.out.println(mce.calculateResolvedType().describe());
    }

**Comment by ghost** [comment]
@jlerbsc Thank you. It works. But I want to test it using the whole project which contains multiple source files and depends on multiple jars.
So how can I set the path (a directory) for `JarTypeSolver ` and `StaticJavaParser`?
I previously set it up using the following, but would have the problem of no response as mentioned above.
```java

String SOURCES_Directory = "/Users/liuyuan/Downloads/dataset/traccar-4.6/src/main/java";
String JARS_Directory = "/Users/liuyuan/Downloads/dataset/traccar-4.6/lib";
Path sourcePath = Paths.get(SOURCES_Directory);
Path jarPath = Paths.get(JARS_Directory);
SymbolSolverCollectionStrategy symbolSolverCollectionStrategy = new SymbolSolverCollectionStrategy();
ProjectRoot projectRoot = symbolSolverCollectionStrategy.collect(sourcePath);
symbolSolverCollectionStrategy.collect(jarPath);
List<SourceRoot> sourceRoots = projectRoot.getSourceRoots();
```

**Comment by jlerbsc** [comment]
Hi @liuyuanOUC, i do not recommend you to use StaticJavaParser (it's mainly for test purpose).
SymbolSolverCollectionStrategy  find all libraries and java source files and configures itself automatically. So this code snippet does the job

    String ROOT_Directory = "/Users/liuyuan/Downloads/dataset/traccar-4.6";
    Path rootPath = Paths.get(ROOT_Directory);
    SymbolSolverCollectionStrategy symbolSolverCollectionStrategy = new SymbolSolverCollectionStrategy();
    ProjectRoot projectRoot = symbolSolverCollectionStrategy.collect(rootPath);
    List<SourceRoot> sourceRoots = projectRoot.getSourceRoots();

But it's possible that you encounter this issue #1771 that is not resolved for the moment.

**Comment by ghost** [comment]
@jlerbsc Thank you for your reply.

I tried your code snippet on Mac (OPENJDK-16 and Javaparser-3.19.0 ) and Windows (JDK15 and Javaparser-3.19.0 ). Unfortunately, it doesn't work because of unable to find jars.

On Mac, it can run. But when it encounters a Type declared in a third-party Jar, it will throw an exception:
```java
Exception in thread "main" java.lang.RuntimeException: javassist.NotFoundException: io.netty.channel.Channel
......
Caused by: javassist.NotFoundException: io.netty.channel.Channel
        at javassist.ClassPool.get(ClassPool.java:430)
        at javassist.bytecode.Descriptor.toCtClass(Descriptor.java:571)
        at javassist.bytecode.Descriptor.getParameterTypes(Descriptor.java:424)
        at javassist.CtBehavior.getParameterTypes(CtBehavior.java:323)
        at com.github.javaparser.symbolsolver.javassistmodel.JavassistMethodDeclaration.getNumberOfParams(JavassistMethodDeclaration.java:141
```

On Windows, it will check for all dependencies immediately on startup, and then throws an exception:
```java
javassist.NotFoundException: F:\dataset\traccar-4.6\._netty-all-4.1.39.Final.jar
```
(I don't know why it adds a prefix '._' for netty-all-4.1.39.Final.jar)

Besides, I checked the `Unpooled ` class and it doesn't implement an interface.

Do you have any ideas or suggestions? Thanks.

**Comment by jlerbsc** [comment]
Considering that the library is configured correctly, i think the NotFoundException is related to this issue # 1771. I'll push a fix today but tests I realized yesterday showed that there were certainly other problems mainly related to performance (I think). We'll see if this happens in your environment. I have already made small improvements in the version under development after analyzing how your sample code works with jvisualvm.

**Comment by ghost** [comment]
@jlerbsc Thanks again~
Hope everything goes well!

**Comment by maartenc** [comment]
I think this issue has been fixed by #3167 as well.
Feel free te reopen it if that is not the case.

**Comment by ghost** [comment]
@maartenc thank you for your effort.
